packages/shared/src/models-hdf5.ts

```diff
@@ -155,6 +157,10 @@ export type Primitive<T extends DType> = T extends NumericType
   ? H5WebComplex
   : unknown;
 
+export type ArrayValue<T extends DType> =
+  | Primitive<T>[]
+  | (T extends NumericType ? TypedArray : never);
```

Comment:
This is the key new type for this PR. `ArrayValue<PrintableType>` resolves to `Primitive<PrintableType>[] | TypedArray`, or `number[] | string[] | ... | TypedArray`.

Once we're ready for providers to return typed arrays, I'll use `ArrayValue<D['type']>` below on line 167.

Reply:
I think we could make the name `ArrayValue` more explicit but that will do.